### PR TITLE
feat: enhances FeatureTracker with spec

### PR DESCRIPTION
**Pull request by cam-garrison**
## Description

This is another PR adding additional functionality to the FeatureTracker objects. This PR fills out the FeatureTrackerSpec with the `Origin` struct which denotes whether the Feature was created by a DSCI or a DSC Component + the name of the DSCI or Component as well as the AppNamespace corresponding with the Feature. 

The goal here is again to enhance usability of the FeatureTrackers outside of just being used to cleanup resources in cluster, giving users/admins more information about resources/features created.

Looks like this: 

```
- apiVersion: features.opendatahub.io/v1
  kind: FeatureTracker
  metadata:
    creationTimestamp: "2024-01-17T20:25:09Z"
    generation: 1
    name: opendatahub-mesh-metrics-collection
    resourceVersion: "78098"
    uid: 2a74ef7d-b1be-4b4f-985a-220978f209b4
  spec:
    appNamespace: opendatahub
    origin:
      name: default
      type: DSCI
  status: {}
- apiVersion: features.opendatahub.io/v1
  kind: FeatureTracker
  metadata:
    creationTimestamp: "2024-01-17T20:31:25Z"
    generation: 1
    name: opendatahub-serverless-serving-deployment
    resourceVersion: "81690"
    uid: 3f4aaff5-e026-4191-8550-942e6c46c9df
  spec:
    appNamespace: opendatahub
    origin:
      name: kserve
      type: Component
  status: {}
```

## How Has This Been Tested?

I've added additional unit tests testing the spec and ensuring it's created correctly as expected. 

I've also tested manually using CRC with operator image `quay.io/cgarriso/opendatahub-operator:incubation-tracker-spec`, creating a DSCI and DSC with kserve enabled, checking that feature trackers are created with the expected specs. 

## Merge criteria:

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: components/kserve/serverless_setup.go [hunk 1]
```diff
@@ -15,55 +17,55 @@ const (
...
-		PostConditions(
-			feature.WaitForPodsToBeReady(knativeServingNamespace),
-		).
-		Load()
-	if err != nil {
-		return err
-	}
-	s.Features = append(s.Features, servingDeployment)
+		servingDeployment, err := feature.CreateFeature("serverless-serving-deployment").
+			For(dscispec, origin).
```
Comment (zdtsw) [inline components/kserve/serverless_setup.go:28]: i must be confused, but should not serverless has DSC from origin? 
(not related to this PR but the old code base) 

e.g Type: featurev1.ComponentType, here should set to use 
`DSCType OwnerType = "DSC"`
not to use `DSCIType OwnerType = "DSCI"`

File: components/kserve/kserve.go [hunk 2]
```diff
@@ -184,9 +185,13 @@ func (k *Kserve) configureServerless(instance *dsciv1.DSCInitializationSpec) err
 	case operatorv1.Managed: // standard workflow to create CR
 		switch instance.ServiceMesh.ManagementState {
 		case operatorv1.Unmanaged, operatorv1.Removed:
-			return fmt.Errorf("ServiceMesh is need to set to 'Managaed' in DSCI CR, it is required by KServe serving field")
+			return fmt.Errorf("ServiceMesh is need to set to 'Managed' in DSCI CR, it is required by KServe serving field")
 		}
-		serverlessInitializer := feature.NewFeaturesInitializer(instance, k.configureServerlessFeatures)
+		origin := featurev1.Origin{
+			Type: featurev1.ComponentType,
+			Name: k.GetComponentName(),
```
Comment (zdtsw) [inline components/kserve/kserve.go:192]: i have been thinking, if one day we have a FT CR which is shared by multiple components.  (similar to odh-model-controller deployed by both kserve and modelmesh) how should we set this Name

File: components/kserve/serverless_setup.go (hunk 1 above)
Comment (zdtsw) [reply components/kserve/serverless_setup.go:28]: ok, i think kinda understand it here. 
it is either DSCI + DSCI CR's name
or  Component + component's name.

File: components/kserve/kserve.go (hunk 2 above)
Comment (bartoszmajsak) [reply components/kserve/kserve.go:192]: That would mean that actual instance of the struct needs to be created in some common base for both. Or we use some suffix for the feature name.

File: components/kserve/serverless_setup.go (hunk 1 above)
Comment (bartoszmajsak) [reply components/kserve/serverless_setup.go:28]: That's right, but it's kinda blurry. For now, maybe we can rework it to not have one `For` for both, as it doesn't make sense. Lemme work with @cam-garrison so we can improve the API and close this PR.

**Comment by cam-garrison** [comment]
@zdtsw Hey, re-requesting review on this as we reworked `For()` to be broken down into `With()` and `DefinedBy()` - let me know if this makes it more readable. 

@israel-hdez - your review on that aspect of the PR would also be helpful, since you've worked with creating Features - again let me know if breaking For() down makes more sense. Thanks!

**Review by bartoszmajsak — APPROVED** [review]
/lgtm

**Review by zdtsw — APPROVED** [review]

**Comment by zdtsw** [comment]
did a live build test, looks fine with the result

File: pkg/feature/builder.go [hunk 3]
```diff
@@ -27,24 +27,44 @@ type featureBuilder struct {
...
+		name: fn.name,
+	}
+}
+
+type featureOrigin struct {
+	spec *v1.DSCInitializationSpec
+	name string
+}
+
+func (fo *featureOrigin) DefinedBy(origin featurev1.Origin) *featureBuilder {
```
Comment (zdtsw) [inline pkg/feature/builder.go:52]: you mind renaming it from `DefinedBy(origin featurev1.Origin)` to `From(source featurev1.Source) `

personally speaking, DeinfedBy and origin are not very easy to understand.

Comment (cam-garrison) [reply pkg/feature/builder.go]: done here! 
1f7c936

**Comment by israel-hdez** [comment]
> @israel-hdez - your review on that aspect of the PR would also be helpful, since you've worked with creating Features - again let me know if breaking For() down makes more sense. Thanks!

Post merge feedback:

I'm OK with the added information and it seems to be useful. Not sure how useful it is going to be for users or admins, but surely it can help with debugging.

What I saw weird is the fact that the `configure${module}Features` funcs are now returning a `func`. On my perception, it could have been left as a plain func. Although, maybe it is just me since I would prefer less indenting.

Now, having said that the added data can be useful, I'm now starting to think that maybe we should, simply, move back to using the `status` field of the `DSCI` and the `DSC` to record everything.

**Pull request merged**